### PR TITLE
complete inference for intersection types

### LINKED ISSUE #36: impossible branches are not ignored
With #34, the following test fails:

~~~erlang
-module(test).

-compile(export_all).

-spec foo2 (a) -> 1; (b) -> 2.
foo2(X) ->
    case X of
        a -> 1;
        b -> 2
    end.
~~~

We now investigate what happens when `foo2` is type-checked against `(a) -> 1`. 
For the `case X of ...` part, the following constraints are generated:

~~~
{"case branch":10:1,
      #{A1@0 => $3, X@0 => $4},
      [],
      [{ccase,
        "case":11:5,
        [{cvar, "var X@0":11:10, X@0, $5},
         {csubty, "exhaustiveness check":11:5, $5, a | b}],
        [{"case branch":12:9,
          #{X@0 => $5 /\ a},
          [],
          [{csubty, "int literal":12:14, 1, $6}],
          $5 /\ a},
         {"case branch":13:9,
          #{X@0 => ($5 /\ not(a)) /\ b},
          [],
          [{csubty, "int literal":13:14, 2, $6}],
          ($5 /\ not(a)) /\ b}]},
       {csubty, "result of case":11:5, $6, $1}],
      $0 /\ ({any()} /\ {any()})}
~~~

Here `$5` is a fresh type variable generated for the scrutiny `X`. From the environment, we get that `$4` (the type of `X@0`) must be `a`.

The contraints above can be broken down as follows:

Constraints for the scrutiny X:

~~~
        [{cvar, "var X@0":11:10, X@0, $5},
         {csubty, "exhaustiveness check":11:5, $5, a | b}],
~~~

These two constraints require that the type of `X` is a subtype of `$5` and the exhaustiveness check requires `$5 <= a | b`. By knowing that `X@0` has type `a`, **this should give us the substitution** `<$5 -> a>`.

Constraints for the first clause:

~~~
        {"case branch":12:9,
          #{X@0 => $5 /\ a},
          [],
          [{csubty, "int literal":12:14, 1, $6}],
          $5 /\ a}
~~~

The env `#{X@0 => $5 /\ a}` refines the type of `X`, `$6` is a fresh type variable for the result type. `$5 /\ a` is the guard type of the clause: only if this type is not bottom, the clause can be taken. Under the substitution `<$5 -> a>`, the guard type becomes `a`, which is not bottom.

Constraints for the second clause:

~~~
         {"case branch":13:9,
          #{X@0 => ($5 /\ not(a)) /\ b},
          [],
          [{csubty, "int literal":13:14, 2, $6}],
          ($5 /\ not(a)) /\ b}]}
~~~

This is very similar to the first clause. The guard type is now `($5 /\ not(a)) /\ b`, which can be simplified to `($5 /\ b)`. Under the substitution `<$5 -> a>`, the guard type becomes `a /\ b`, which is bottom. This reflects the fact that the second clause is impossible if the argument type is `a`.

The problem arises during constraint simplification. Here are the relevant parts from the log:

~~~
[DEBUG 2023-09-18 09:41:43.246 constr_simp.erl:101] Checking which branches of case at test.erl:11:5 should be ignored.
Env: {A1@0 => atom() /\ (mu8 | not(mu8) /\ a) |
         integer() /\ mu8 |
         tuple() /\ (mu8 /\ tuple()) |
         fun() /\ (mu8 /\ fun()),
 X@0 => atom() /\ (mu7 | not(mu7) /\ a) |
        integer() /\ mu7 |
        tuple() /\ (mu7 /\ tuple()) |
        fun() /\ (mu7 /\ fun())}
Fixed tyvars: [mu7,mu8]
Constraints for scrutiny:
{[{csubty, "exhaustiveness check":11:5, $5, a | b},
  {csubty,
   "var X@0":11:10,
   atom() /\ (mu7 | not(mu7) /\ a) |
   integer() /\ mu7 |
   tuple() /\ (mu7 /\ tuple()) |
   fun() /\ (mu7 /\ fun()),
   $5}]}
[DEBUG 2023-09-18 09:41:43.246 constr_simp.erl:111] Unique substitution:
{$5 => atom() /\ (mu15 /\ (a | b) | not(mu15) /\ a),
 mu7 => atom() /\ (mu14 /\ (mu15 /\ (a | b) | not(mu15) /\ a))}
~~~

We see that `X@0` does not get the type `a` but a much more complicated type:

~~~
        atom() /\ (mu7 | not(mu7) /\ a) |
        integer() /\ mu7 |
        tuple() /\ (mu7 /\ tuple()) |
        fun() /\ (mu7 /\ fun())
~~~

We can slightly simplify this type

~~~
        atom() /\ mu7 | 
        not(mu7) /\ a) |
        integer() /\ mu7 |
        tuple() /\ mu7 |
        fun() /\ mu7
~~~

But still, it's not the same as `a` (what we would expect). Hence, we get the following substitution:

~~~
<$5 -> mu15 /\ (a | b) | not(mu15) /\ a>
~~~

But under this substitution, the type `$5 /\ b` is not bottom, so the second branch ist not ignored.

### PR DESCRIPTION
**Pull request by skogsbaer**
Fixes #36 and #56

There are two outstand problems with intersection types, tracked in #103 

I think we should merge this PR regardless of these open problems.

* Support for Erlang 24 is dropped

### CODE REVIEW

**Comment by skogsbaer** [comment]
@albsch now also merge with main.

File: test/tycheck_simple_tests.erl [hunk 1]
```diff
@@ -7,9 +7,9 @@
 
 -spec check_ok_fun(string(), symtab:t(), ast:fun_decl(), ast:ty_scheme()) -> ok.
 check_ok_fun(Filename, Tab, Decl = {function, L, Name, Arity, _}, Ty) ->
-    Ctx = typing:new_ctx(Tab, error),
+    Ctx = typing:new_ctx(Tab, error), % FIXME: perform sanity check!
```
Comment (albsch) [inline test/tycheck_simple_tests.erl:10]: What does this refer to?

File: test/tycheck_simple_tests.erl [hunk 2]
```diff
@@ -133,10 +133,8 @@ should_run(Name, {exclude,Set}) -> not sets:is_element(Name, Set).
 simple_test_() ->
   % The following functions are currently excluded from being tested.
   WhatNot = [
-    % buggy, see #36
-    "fun_local_03",
-    "fun_local_07",
-    "fun_local_07_fail"
+    % Redundancy check for lists is not powerful enough
```
Comment (albsch) [inline test/tycheck_simple_tests.erl:136]: Is this tracked somewhere?

**Review by albsch — CHANGES_REQUESTED** [review]

File: test/tycheck_simple_tests.erl (hunk 2 above)
Comment (skogsbaer) [reply test/tycheck_simple_tests.erl:136]: #108

File: test/tycheck_simple_tests.erl (hunk 1 above)
Comment (skogsbaer) [reply test/tycheck_simple_tests.erl:10]: Fixed

**Comment by skogsbaer** [comment]
@albsch both issues addressed.

**Review by albsch — APPROVED** [review]

**Pull request merged**